### DISCUSSION TITLE
Error Failed to load PDF document.

### DISCUSSION THREAD

**Discussion by Teddy937** [spatie/browsershot#866] [Q&A]
Am getting this error on ubuntu server when trying to display pdf generate by spatie pdf , Locally everything works fine

**Comment by hailwood** [comment]
@Teddy937 out of curiosity, if you save the file to disk instead of downloading it/rendering it inline - does the saved pdf open?

**Reply by Teddy937** [reply]
No, The file is not open @hailwood

**Reply by Teddy937** [reply]
[Teddy_1723333879.pdf](https://github.com/user-attachments/files/16576847/Teddy_1723333879.pdf)
Here is a sample file

**Reply by hailwood** [reply]
What code did you use to generate the file?

**Reply by Teddy937** [reply]
Actually the code works fine locally and generates pdf locally, On Ubuntu server is where its behaving that way

```public function process_pdf($username, $template)
    {
        try {
            $expert = User::where('username', $username)->first();
            $profile = "https://expert195.com/profile/$username";
            if (!$expert->profile_url) {
                $expert->update(['profile_url' => $profile]);
                $expert = User::where('username', $username)->first();
            }

            $name = $expert->username . '_' . time();
            $location = $expert->location;
            if (!$location) {
                $workArea = Expert_location::where('user_id', $expert->id)->first();
                if ($workArea) {
                    $location = $workArea->location_address;
                } else {
                    $location = 'N/A';
                }
            }
            $sample = $template == 'default' ? 'pdfs.default' : 'pdfs.expert_card';
            return pdf()
                ->view($sample, compact('expert', 'location'))
                ->name($name);
        } catch (\Throwable $th) {
            return response()->json(['message' => $th->getMessage()], Response::HTTP_INTERNAL_SERVER_ERROR);
        }
    }```

**Reply by hailwood** [reply]
Sweet, can you please try doing 

```php
return pdf()
->view($sample, compact('expert', 'location'))
->save('/some/location/file.pdf'); // e.g. ->save(storage_path('invoice.pdf'));
```

And then find and open the saved file.
I think you might be running into the same bug as me, if so the file will open when you save it like that.

**Reply by bluesheep100** [reply]
I've also experienced this issue in a local Sail environment. Downgrading to Puppeteer 22.15 (latest version before 23.0.0) fixes the issue.
I attempted to recreate the problem with just Puppeteer in a pure Nodejs environment, with the same versions, but I can't seem to make it happen.
Clearly, there is some kind of encoding mishap that makes it return a corrupt PDF, but it doesn't seem to be directly caused by Puppeteer.

**Comment by bluesheep100** [comment]
I receive the same error, using this code:
```php
$pdf = BrowserShot::url($url)
    ->noSandbox()
    ->setChromePath('/usr/bin/google-chrome')
    ->format('a4');

return Response::make($pdf->pdf(), headers: [
    'Content-Type' => 'application/pdf',
    'Content-Disposition' => "inline; filename=\"$filename\"",
]);
```

If instead of returning a response I use `$pdf->save($filename);` and open the file in a PDF reader, or the browser, it works fine.

**Reply by hailwood** [reply]
Yeah, so from my investigations it seems that this line https://github.com/spatie/browsershot/blob/main/bin/browser.cjs#L52 isn't returning a base64 string, instead its returning what looks like a uint array.

Then we're base64 decoding it here https://github.com/spatie/browsershot/blob/main/src/Browsershot.php#L677 which results in total garbage output.

Saving the pdf to disk works because the node process is just writing the buffer and then we don't attempt to decode it.

Unfortunately, I have no idea how to fix it, or whether the issue is a puppeteer issue, a chrome/chromium version issue, or just Browsershot doing something wrong.

@freekmurze is this of any help to you?

**Reply by bluesheep100** [reply]
Okay, this explains things. When replicating this in puppeteer in a more isolated environment, I was getting the UInt8Array as well, and it did seem odd. In that case, it's definitely from Puppeteer's side of things, so I guess all we can do is wait for a fix.

**Comment by JeppeKnockaert** [comment]
The output from Puppeteer was changed in v23: https://github.com/puppeteer/puppeteer/pull/12823 and https://github.com/puppeteer/puppeteer/releases/tag/puppeteer-core-v23.0.0

I've created a PR to fix the issue: https://github.com/spatie/browsershot/pull/869

**Comment by tasso85** [comment]
I see that there is a new release which says should solve this problem when using Puppeteer 23.x, however, after upgrading Browsershot, I am still getting the error when trying to display the inline PDF, without any exception being logged.

As soon as I revert to Puppeteer 22, it starts working again.

Did I maybe understood wrong, that latest 4.2.1 release fixes the issue?

**Reply by PHLAK** [reply]
I believe 4.2.1 revered the changes from 4.2.0, so 4.2.1 is effectively the same as 4.1.3 (which doesn't work with Puppeteer 23).

**Reply by tasso85** [reply]
Oh, so my mistake then, thanks for clarifying.

**Comment by Teddy937** [comment]
So after downgrading back to version 22.15.0 as @bluesheep100  mentioned, It still did not work. After giving puppeteer appropriate permissions  to access the .cache/puppeteer directory which I copied to root folder of my project and pointed to it using .puppeteerrc.cjs as described here: [here](https://pptr.dev/guides/configuration#changing-the-default-cache-directory).
These are the commands I ran to give puppeteer permissions: ```sudo chown -R www-data:www-data /var/www/.cache/puppeteer``` and ```sudo chmod -R 755 /var/www/.cache/puppeteer```
